cmd/gpu_nfdhook/labeler.go

```diff
@@ -171,33 +171,70 @@ func (l *labeler) createCapabilityLabels(cardNum string, numTiles uint64) {
 	}
 	defer file.Close()
 
-	// define strings to search from the file, and the actions to take in order to create labels from those strings (as funcs)
+	gen := ""
+	media := ""
+	graphics := ""
+	// define string prefixes to search from the file, and the actions to take in order to create labels from those strings (as funcs)
 	searchStringActionMap := map[string]func(string){
-		"platform: ": func(platformName string) {
+		"platform:": func(platformName string) {
 			l.labels.addNumericLabel(labelNamespace+"platform_"+platformName+".count", 1)
 			l.labels[labelNamespace+"platform_"+platformName+".tiles"] = strconv.FormatInt(int64(numTiles), 10)
 			l.labels[labelNamespace+"platform_"+platformName+".present"] = "true"
 		},
-		"gen: ": func(genName string) {
-			l.labels[labelNamespace+"platform_gen"] = genName
+		// there's also display block version, but that's not relevant
+		"media version:": func(version string) {
+			l.labels[labelNamespace+"media_version"] = version
+			media = version
+		},
+		"graphics version:": func(version string) {
+			l.labels[labelNamespace+"graphics_version"] = version
+			graphics = version
+		},
+		"gen:": func(version string) {
+			l.labels[labelNamespace+"platform_gen"] = version
+			gen = version
 		},
 	}
 
 	// Finally, read the file, and try to find the matches. Perform actions and reduce the search map size as we proceed. Return at 0 size.
 	scanner := bufio.NewScanner(file)
+scanning:
 	for scanner.Scan() {
-		for searchString, action := range searchStringActionMap {
-			var stringValue string
-			n, _ := fmt.Sscanf(scanner.Text(), searchString+"%s", &stringValue)
-			if n > 0 {
-				action(stringValue)
-				delete(searchStringActionMap, searchString)
-				if len(searchStringActionMap) == 0 {
-					return
-				}
-				break
+		line := scanner.Text()
+		for prefix, action := range searchStringActionMap {
+			if !strings.HasPrefix(line, prefix) {
+				continue
+			}
+			fields := strings.Split(line, ": ")
+			if len(fields) == 2 {
+				action(fields[1])
+			} else {
+				klog.Warningf("invalid '%s' line format: '%s'", file.Name(), line)
+			}
+			delete(searchStringActionMap, prefix)
```

Comment:
Mutating a collection while iterating over it feels slightly unsafe. Can we replace it with a counter compared with initial `len(searchStringActionMap)`?

Reply:
It is safe. https://stackoverflow.com/questions/23229975/is-it-safe-to-remove-selected-keys-from-map-within-a-range-loop

Reply:
It's from Ukri's original code.  The spec: https://go.dev/ref/spec#For_statements

Says just that: "If a map entry that has not yet been reached is removed during iteration, the corresponding iteration value will not be produced."

But according to this, adding and deleting map entries during range would be safe: https://github.com/golang/go/issues/9926

I think this is because "range" provides items by value (i.e. copy), not by reference.

But propose a quick code example how you'd like it to go, so that we can compare!

Reply:
I know it's safe. It's just my tiny brain keeps forgetting it and I have to google every time I see the pattern. :)
An alternative code could probably be this
```go
        scanner := bufio.NewScanner(file)
        for scanner.Scan() {
                line := scanner.Text()
                foundPrefix := ""
                for prefix, action := range searchStringActionMap {
                        if strings.HasPrefix(line, prefix) {
                                fields := strings.Split(line, ": ")
                                if len(fields) == 2 {
                                        action(fields[1])
                                } else {
                                        klog.Warningf("invalid '%s' line format: '%s'", file.Name(), line)
                                }
                                foundPrefix = prefix
                                break
                        }
                }

                delete(searchStringActionMap, foundPrefix)
                if len(searchStringActionMap) == 0 {
                        break
                }
        }

```

Reply:
Looks OK, but is extra delete + len + comparison operation per capability file line, as map with _current setup_ never goes empty.  Might as well drop then the len + break completely (at least until "gen" support goes away completely).  It's probably not much extra compared to range with its copy semantics though.

Any strong feelings from others?

PS. Delete within range map is given as an actual example even in Effective Go: https://go.dev/doc/effective_go#for

Reply:
Alright, let's keep delete within the loop.